src/Sarif.UnitTests/Visitors/InsertOptionalDataVisitorTests.cs

```diff
@@ -0,0 +1,157 @@
+
+            string expectedFileName = inputFileName + optionsNameSuffix + ".sarif";
+            string actualFileName = inputFileName + optionsNameSuffix + ".actual.sarif";
+            inputFileName = inputFileName + ".sarif";
+
+            expectedFileName = Path.Combine(testDirectory, expectedFileName);
+            actualFileName = Path.Combine(testDirectory, actualFileName);
+            inputFileName = Path.Combine(testDirectory, inputFileName);
+
+            File.Exists(inputFileName).Should().BeTrue();
+
+            SarifLog actualLog;
+
+            JsonSerializerSettings settings = new JsonSerializerSettings()
+            {
+                ContractResolver = SarifContractResolver.Instance,
+                Formatting = Formatting.Indented
+            };
+
+            try
+            {
+                actualLog = JsonConvert.DeserializeObject<SarifLog>(File.ReadAllText(inputFileName), settings);
+
+                var visitor = new InsertOptionalDataVisitor(optionallyEmittedData);
+                visitor.Visit(actualLog.Runs[0]);
+            }
+            catch (Exception ex)
+            {
+                sb.AppendLine(string.Format(CultureInfo.InvariantCulture, "Unhandled exception processing input '{0}' with the following options: '{1}'.", inputFileName, optionallyEmittedData));
+                sb.AppendLine(ex.ToString());
+                ValidateResults(sb.ToString());
+                return;
+            }
+
+            string expectedSarif = File.Exists(expectedFileName) ? File.ReadAllText(expectedFileName) : null;
+            string actualSarif = JsonConvert.SerializeObject(actualLog, settings);
+
+            if (!AreEquivalentSarifLogs(actualSarif, expectedSarif))
+            {
+                if (s_rebaseline)
+                {
+                    // We rewrite to test output directory. This allows subsequent tests to 
+                    // pass without requiring a rebuild that recopies SARIF test files
+                    File.WriteAllText(expectedFileName, actualSarif);
+
+                    string subdirectory = Path.GetFileName(testDirectory);
+                    string productTestDirectory = GetProductTestDataDirectory(subdirectory);
+                    expectedFileName = Path.GetFileName(expectedFileName);
+                    expectedFileName = Path.Combine(productTestDirectory, expectedFileName);
+
+                    // We also rewrite the checked in test baselines
+                    File.WriteAllText(expectedFileName, actualSarif);
+                }
+                else
+                {
+                    File.WriteAllText(actualFileName, actualSarif);
+
+                    string errorMessage = "Expanding optional data for input '{0}' produced unexpected results for the following options: '{1}'.";
+                    sb.AppendLine(string.Format(CultureInfo.CurrentCulture, errorMessage, inputFileName, optionallyEmittedData));
```

Comment:
>sb.AppendLine(string.Format(CultureInfo.CurrentCulture [](start = 20, length = 54)

`StringBuilder.AppendFormat(IFormatProvider, ...)` is the overload you need.

Reply:
well, at the expense of having to figure out how to handle the newline addition, as embedded chars or use of Environment.NewLine (which increases line length)

---
In reply to: [205988006](https://github.com/Microsoft/sarif-sdk/pull/974#discussion_r205988006) [](ancestors = 205988006)